state/store.go

```diff
@@ -253,13 +254,41 @@ func PruneStates(db dbm.DB, from int64, to int64) error {
 	return nil
 }
 
-// ABCIResponsesResultsHash returns the merkle hash of the deliverTxs within ABCIResponses
-func ABCIResponsesResultsHash(ar tmstate.ABCIResponses) []byte {
+//------------------------------------------------------------------------
+
+// ABCIResponsesResultsHash returns the root hash of a Merkle tree with 3 leafs:
+//   1) proto encoded ResponseBeginBlock.Events
+//   2) root hash of a Merkle tree of ResponseDeliverTx responses (see ABCIResults.Hash)
+//   3) proto encoded ResponseEndBlock.Events
+//
+// See merkle.SimpleHashFromByteSlices
+func ABCIResponsesResultsHash(ar *tmstate.ABCIResponses) []byte {
+	// proto-encode BeginBlock events.
+	bbeBytes, err := proto.Marshal(&abci.ResponseBeginBlock{
+		Events: ar.BeginBlock.Events,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	// Build a Merkle tree of amino-encoded DeliverTx results and get a hash.
 	results := types.NewResults(ar.DeliverTxs)
-	return results.Hash()
+
+	// proto-encode EndBlock events.
+	ebeBytes, err := proto.Marshal(&abci.ResponseEndBlock{
```

Comment:
not sure if it's the best way, but kinda works

Reply:
this looks good.

Reply:
you mean `X.Marshal()`?

Reply:
yes